src/render/filters.rs

```diff
@@ -501,6 +503,99 @@ impl ResolveFilter for WordcountFilter {
     }
 }
 
+/// A word-wrap function that preserves existing line breaks.
+/// Expects that existing line breaks are posix newlines.
+///
+/// Preserve all white space except added line breaks consume the space on
+/// which they break the line.
+///
+/// Don't wrap long words, thus the output text may have lines longer than ``width``.
+fn wordwrap(text: &str, width: usize) -> String {
+    let mut result = String::with_capacity(text.len());
+
+    for (i, line) in text.split('\n').enumerate() {
+        if i > 0 {
+            result.push('\n');
+        }
+
+        if line.is_empty() {
+            continue;
+        }
+
+        let leading_whitespace = line.chars().take_while(|c| c.is_whitespace()).count();
+        let (indent, trimmed_line) = line.split_at(leading_whitespace);
+
+        let mut words = trimmed_line.split_whitespace();
+
+        let Some(first_word) = words.next() else {
+            // Line contains only whitespace - preserve it
+            result.push_str(line);
+            continue;
+        };
+
+        result.push_str(indent);
+        result.push_str(first_word);
+
+        let mut current_len = leading_whitespace + first_word.len();
+        for word in words {
+            let word_len = word.len();
+
+            if current_len + word_len < width {
+                result.push(' ');
+                current_len += 1 + word_len;
+            } else {
+                result.push('\n');
+                current_len = word_len;
+            }
+            result.push_str(word);
+        }
+    }
+
+    result
+}
+
+impl ResolveFilter for WordwrapFilter {
+    fn resolve<'t, 'py>(
+        &self,
+        variable: Option<Content<'t, 'py>>,
+        py: Python<'py>,
+        template: TemplateString<'t>,
+        context: &mut Context,
+    ) -> ResolveResult<'t, 'py> {
+        let Some(content) = variable else {
+            return Ok(Some("".as_content()));
+        };
+
+        let text = content.resolve_string(context)?;
+
+        let arg = self
+            .argument
+            .resolve(py, template, context, ResolveFailures::Raise)?
+            .expect("missing argument in context should already have raised");
+
+        // Check for negative values before converting to usize
+        if let Some(bigint) = arg.to_bigint()
+            && let Some(n) = bigint.to_isize()
+            && n <= 0
+        {
+            return Err(pyo3::exceptions::PyValueError::new_err(format!(
+                "invalid width {n} (must be > 0)"
+            ))
+            .annotate(py, self.argument.at, "width", template)
+            .into());
+        }
+
+        let width = match arg.resolve_usize(self.argument.at) {
+            Ok(w) => w,
+            Err(PyRenderError::RenderError(RenderError::OverflowError { .. })) => usize::MAX,
+            Err(e) => return Err(e),
+        };
```

Comment:
I suspect we can optimise this logic a bit more, but it's probably not worth it.